File: Exercise/Very Hard/02. Shortest Path Finder/Code/Shortest Path Finder.cs
using System;
using System.Collections.Generic;

class Program {
    class Node : IComparable<Node> {
        public int Distance;
        public int Vertex;
        public List<int> Path;

        public int CompareTo(Node other) {
            return Distance.CompareTo(other.Distance);
        }
    }

    static List<int> ShortestPath(int n, int[,] distances, int start, int end) {
        var minHeap = new SortedSet<Node>(Comparer<Node>.Create((a, b) => 
            a.Distance == b.Distance ? a.Vertex.CompareTo(b.Vertex) : a.Distance.CompareTo(b.Distance)));
        minHeap.Add(new Node { Distance = 0, Vertex = start, Path = new List<int>() });

        while (minHeap.Count > 0) {
            var current = minHeap.Min;
            minHeap.Remove(current);

            if (current.Vertex == end) {
                current.Path.Add(current.Vertex);
                return current.Path;
            }

            for (int neighbor = 0; neighbor < n; neighbor++) {
                if (distances[current.Vertex, neighbor] != int.MaxValue && neighbor != current.Vertex) {
                    var newPath = new List<int>(current.Path) { current.Vertex };
                    minHeap.Add(new Node {
                        Distance = current.Distance + distances[current.Vertex, neighbor],
                        Vertex = neighbor,
                        Path = newPath
                    });
                }
            }
        }

        return new List<int>();
    }

    static void Main() {
        int n = 5;
        int[,] distances = {
            { 0, 3, 1, int.MaxValue, int.MaxValue },
            { 3, 0, 7, 5, 1 },
            { 1, 7, 0, 2, 3 },
            { int.MaxValue, 5, 2, 0, 7 },
            { int.MaxValue, 1, 3, 7, 0 }
        };

        int start = 0;
        int end = 4;

        var result = ShortestPath(n, distances, start, end);
        Console.WriteLine(string.Join(" ", result));
    }
}

File: Exercise/Very Hard/02. Shortest Path Finder/Code/Shortest Path Finder.java
import java.util.*;

class Main {
    static class Node implements Comparable<Node> {
        int distance;
        int vertex;
        List<Integer> path;

        Node(int distance, int vertex, List<Integer> path) {
            this.distance = distance;
            this.vertex = vertex;
            this.path = path;
        }

        @Override
        public int compareTo(Node other) {
            return Integer.compare(this.distance, other.distance);
        }
    }

    public static List<Integer> shortestPath(int n, int[][] distances, int start, int end) {
        PriorityQueue<Node> minHeap = new PriorityQueue<>();
        minHeap.add(new Node(0, start, new ArrayList<>()));

        while (!minHeap.isEmpty()) {
            Node current = minHeap.poll();

            if (current.vertex == end) {
                current.path.add(current.vertex);
                return current.path;
            }

            for (int neighbor = 0; neighbor < n; neighbor++) {
                if (distances[current.vertex][neighbor] != Integer.MAX_VALUE && neighbor != current.vertex) {
                    List<Integer> newPath = new ArrayList<>(current.path);
                    newPath.add(current.vertex);
                    minHeap.add(new Node(current.distance + distances[current.vertex][neighbor], neighbor, newPath));
                }
            }
        }

        return new ArrayList<>();
    }

    public static void main(String[] args) {
        int n = 5;
        int[][] distances = {
            {0, 3, 1, Integer.MAX_VALUE, Integer.MAX_VALUE},
            {3, 0, 7, 5, 1},
            {1, 7, 0, 2, 3},
            {Integer.MAX_VALUE, 5, 2, 0, 7},
            {Integer.MAX_VALUE, 1, 3, 7, 0}
        };

        int start = 0;
        int end = 4;

        List<Integer> result = shortestPath(n, distances, start, end);
        for (int vertex : result) {
            System.out.print(vertex + " ");
        }
        System.out.println();
    }
}




